0012-BIRS-distributed-scheduler.md

```diff
@@ -0,0 +1,463 @@
+    string ttl = 5;
+}
+
+// rpc ScheduleJob(ScheduleJobRequest) returns (google.protobuf.Empty) {}
+
+// ScheduleJobRequest is the message to create/schedule the job.
+message ScheduleJobRequest {
+    // The job details.
+    Job job = 1;
+}
+
+// rpc DeleteJob(DeleteJobRequest) returns (google.protobuf.Empty) {}
+
+// DeleteJobRequest is the message to delete the job by name.
+message DeleteJobRequest {
+    // The name of the job.
+    string name = 1;
+}
+
+// rpc GetJob(GetJobRequest) returns (GetJobResponse) {}
+
+// GetJobRequest is the message to get the job by name.
+message GetJobRequest {
+    // The name of the job.
+    string name = 1;
+}
+
+// GetJobResponse is the response message to convey the job.
+message GetJobResponse {
+    // The job details.
+    Job job = 1;
+}
+
+// rpc ListJobs(ListJobsRequest) returns (ListJobsResponse) {}
+
+// ListJobsRequest is the message to list jobs by app_id.
+message ListJobsRequest {
+    // The id of the application (app_id) for which to list jobs.
+    string app_id = 1;
+}
+
+// ListJobsResponse is the response message to convey the list of jobs.
+message ListJobsResponse {
+    // List of jobs that match the request criteria.
+    repeated Job jobs = 1;
+}
+```
+
+###### Daprd Sidecar to Scheduler Service APIs
+
+For the daprd sidecar to Scheduler Service communication, 
+![Scheduler APIs (SideCar Facing)](./resources/0004-BIRS-distributed-scheduler/sidecarToSchedulerComm.png)
+
+We will use the same exact protos from the Public Dapr API, but inside a **new**: `dapr/proto/scheduler/scheduler.proto`. However, the one difference will be in the `Job message`, where we will add:
+
+```proto
+    // Namespace for jobs to allow for multi tenancy.
+    string namespace = 5;
+
+    // The metadata associated with the job.
+    // The sidecar will create the unique `key` for storing data in the state store and pass the generated `key` along to the scheduler service for data lookup upon ‘trigger’ time later on. 
+    // The sidecar will also add metadata in order to know whether this job is registered for an actor. This is needed, as the routing mechanism for actors is different for the callback.
+    map<string,string> metadata = 6;
+```
+
+
+To allow for the daprd sidecar to register itself with the Scheduler Service upon startup, there will be a new `register` API in `dapr/proto/scheduler/scheduler.proto` with the following:
+
+```proto
+// This will allow for communication between the daprd sidecar and the Scheduler Service. (daprd sidecar -> Scheduler Service)
+service Scheduler {
+
+    ...<same user-facing protos defined above>...
+
+    // Allow for the daprd sidecar to register with a control plane Scheduler Service
+    rpc Register(RegisterRequest) returns (google.protobuf.Empty) {}
+}
+
+// RegisterRequest is the message to register a daprd sidecar with a control plane service. 
+message RegisterRequest {
```

Comment:
I am strongly against registering daprds as a one shot requests, and saving the _hostname_ of the registered app.

- There is no strong guarentee and will often be the case that a daprd will terminate without de-registering, making the record stale.
- It can be expected that daprd's hostnames change over their lifecyles in cloud native environments, also making their records broken.
- We already have an app ID & namespace name resolver in Dapr. This would duplicate that system.
- The scheduler is going to have to consistently ping registered daprds for their health causing non-trivial amounts of traffic and resource usage. Each health check to, say, 1000s daprds is going through the entire network stack every time (DNS & TCP & TLS & close).
- Managing Job request triggers -> app ID & hostnames and keeping those records accurate is a hard problem in of itself.

There is precedent in dapr today for long running streams of active daprds to control plane components with placement and operator. I think it makes sense to continue with that approach.

Reply:
Upon further implementation exploration, we decided it was best to move forward with streaming. The proposal has been updated to reflect current implementation details